### PR TITLE
Fix unsoundess in Resource Leak Checker related to owning fields and EnsuresCalledMethods

### LINKED ISSUE #4838: Resource Leak Checker's rules for owning fields rely on a stronger contract for @EnsuresCalledMethods than the actual contract
The contract of the `@EnsuresCalledMethods` annotation is that it guarantees that the given method(s) are called on the given expression(s) *if the method exits normally*. The rules used by the Resource Leak Checker, however, assume that the given methods are called on all paths. This was first noticed by @msridhar in https://github.com/typetools/checker-framework/pull/4808/files#r673287910.

As a concrete example, consider the following program:

```
import java.io.IOException;
import java.io.InputStream;
import java.io.Closeable;
import org.checkerframework.checker.mustcall.qual.Owning;
import org.checkerframework.checker.calledmethods.qual.EnsuresCalledMethods;

class ReplicaInputStreams implements Closeable {

  private final @Owning InputStream in1;
  private final @Owning InputStream in2;

  public ReplicaInputStreams(
      @Owning InputStream i1, @Owning InputStream i2) {
    this.in1 = i1;
    this.in2 = i2;
  }

  @Override
  @EnsuresCalledMethods(
      value = {"this.in1", "this.in2"},
      methods = {"close"})
  public void close() throws IOException {
    in1.close();
    in2.close();
  }
}
```

Currently, the Resource Leak Checker incorrectly verifies this program (i.e. running `javac -processor resourceleak ReplicaInputStreams.java` does not issue any errors). I would expect some kind of error to be issued about this implementation of `close()`, because it does not guarantee that `in2.close()` is called on all paths - there is an exceptional exit path from `in1.close()`.

Confusingly, a similar example I wrote up **does** issue an error:
```
import java.net.Socket;
import java.io.IOException;
import org.checkerframework.checker.mustcall.qual.*;
import org.checkerframework.checker.calledmethods.qual.*;

@MustCall("dispose")
class TwoResourcesECM {
  @Owning Socket s1, s2;

  @EnsuresCalledMethods(value = {"this.s1", "this.s2"}, methods={"close"})
  public void dispose() throws IOException {
    s1.close();
    s2.close();
  }
}
```

In this nearly identical example, the Called Methods Checker rejects `dispose()`, issuing the following error:
```
!?rlc-twoclose-soundness ~/jsr308/checker-framework> checker/bin/javac -processor resourceleak checker/tests/resourceleak/TwoResourcesECM.java    
checker/tests/resourceleak/TwoResourcesECM.java:14: error: [contracts.postcondition] postcondition of dispose is not satisfied.
  public void dispose() throws IOException {
              ^
  found   : no information about this.s1
  required: this.s1 is @CalledMethods("close")
1 error
exit 1
```

### PR DESCRIPTION
**Pull request by kelloggm**
Fixes #4838.

Merge after #4867.

This is a replacement for #4839, based on the comments on it from @msridhar. It relies on #4867 to resolve the false positives described in the comment chain on #4839.

### CODE REVIEW

**Comment by msridhar** [comment]
This looks pretty clean!  The change looks good to me.  Do we have a test where the finalizer does a `try-finally` and we are able to still verify it?

**Comment by kelloggm** [comment]
> Do we have a test where the finalizer does a try-finally and we are able to still verify it?

We did not, but I just added one as `ReplicaInputStreams2.java`.

File: checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java [hunk 1]
```diff
@@ -80,6 +86,66 @@ public Void visitMethod(MethodTree node, Void p) {
...
+            .getAnnotatedType(containingClass)
+            .getAnnotationInHierarchy(mustCallAnnotatedTypeFactory.TOP);
+    List<String> mcValues =
+        AnnotationUtils.getElementValueArray(
+            mcAnno, mustCallAnnotatedTypeFactory.getMustCallValueElement(), String.class);
+    String methodName = elt.getSimpleName().toString();
+    if (!mcValues.contains(methodName)) {
+      // Not a destructor, just a method with an ECM annotation. No further checking to do.
+      return;
+    }
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/resourceleak/ResourceLeakVisitor.java:120]: Can we extract this code to a new method that returns a boolean?  Something like `isDestructorForEnclosingClass`?

**Review by msridhar — CHANGES_REQUESTED** [review]
Mostly LGTM; just one style comment

File: checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java [hunk 2]
```diff
@@ -20,21 +22,31 @@
   protected CalledMethodsAnalysis(
       BaseTypeChecker checker, CalledMethodsAnnotatedTypeFactory factory) {
     super(checker, factory);
+    // Use the Nullness Checker.
+    ignoredExceptionTypes.add("java.lang.NullPointerException");
+    // Ignore run-time errors, which cannot be predicted statically. Doing so is unsound
+    // in the sense that they could still occur - e.g., the program could run out of memory -
+    // but if they did, the checker's results would be useless anyway.
+    ignoredExceptionTypes.add("java.lang.Error");
+    ignoredExceptionTypes.add("java.lang.RuntimeException");
```
Comment (msridhar) [inline checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java:31]: Why are we adding these exception types in this PR?  Do the tests not pass without it?

**Review by msridhar — COMMENTED** [review]
Sorry, one more question I forgot to ask before

File: checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java (hunk 2 above)
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java:31]: The tests do not pass without it.

In particular, we get these three false positives:
```
    3 unexpected diagnostics were found
    CheckFields.java:68: error: (destructor.exceptional.postcondition)
    CreatesMustCallForOverride2.java:38: error: (destructor.exceptional.postcondition)
    CreatesMustCallForOverride2.java:56: error: (destructor.exceptional.postcondition)
```

For reference, here's the relevant lines of `CheckFields`:
```
    @EnsuresCalledMethods(
        value = {"this.finalOwningFoo", "this.owningFoo"},
        methods = {"a"})
    void b() {
      this.finalOwningFoo.a();
      this.finalOwningFoo.c();
      this.owningFoo.a();
    }
  }
```

Comment (msridhar) [reply checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java:31]: After this lands we may want to think about another manual update like #4892

**Review by msridhar — APPROVED** [review]

File: checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java (hunk 2 above)
Comment (kelloggm) [reply checker/src/main/java/org/checkerframework/checker/calledmethods/CalledMethodsAnalysis.java:31]: I'm not sure that's really necessary - that run-time errors don't invalidate CalledMethods types follows from the idea that annotations (and therefore types) specify *normal behavior*, and a run-time exception is definitely not normal behavior.

**Pull request merged**